examples/SimilarityLearning/mnist-embeddings.py

```diff
@@ -0,0 +1,319 @@
+            a = self.rng.randint(len(self.data_dict[pick_label]))
+            p = self.rng.randint(len(self.data_dict[pick_label]))
+            n = self.rng.randint(len(self.data_dict[pick_other]))
+
+            a = np.reshape(self.data_dict[pick_label][a], [28, 28]).astype(np.float32)
+            p = np.reshape(self.data_dict[pick_label][p], [28, 28]).astype(np.float32)
+            n = np.reshape(self.data_dict[pick_other][n], [28, 28]).astype(np.float32)
+
+            yield [a, p, n]
+
+
+class TripletModel(EmbeddingModel):
+    @staticmethod
+    def get_data():
+        # return [digit_A, digit_B, similar?]
+        ds = MnistTriplets('train')
+        ds = BatchData(ds, 128 // 3)
+        return ds
+
+    def _get_input_vars(self):
+        return [InputVar(tf.float32, (None, 28, 28), 'input'),
+                InputVar(tf.float32, (None, 28, 28), 'input_p'),
+                InputVar(tf.float32, (None, 28, 28), 'input_n')]
+
+    def loss(self, a, p, n):
+        return symbf.triplet_loss(a, p, n, 5., extra=True)
+
+    def _build_graph(self, input_vars):
+        # get inputs
+        a, p, n = input_vars
+        # embed them
+        a, p, n = self.embed([a, p, n])
+        # tag them, just for inference later on
+        a = tf.identity(a, name="emb")
+
+        # compute the actual loss
+        cost, pos_dist, neg_dist = self.loss(a, p, n)
+        self.cost = tf.identity(cost, name="cost")
+
+        # track these values during training
+        tf.summary.scalar('pos_dist', pos_dist)
+        tf.summary.scalar('neg_dist', neg_dist)
+
+
+class SoftTripletModel(TripletModel):
+    def loss(self, a, p, n):
+        return symbf.soft_triplet_loss(a, p, n)
+
+
+def get_config(model):
+    logger.auto_set_dir()
+
+    dataset = model.get_data()
+    step_per_epoch = dataset.size()
+
+    lr = symbolic_functions.get_scalar_var('learning_rate', 1e-4, summary=True)
+
+    return TrainConfig(
+        dataflow=dataset,
+        optimizer=tf.train.GradientDescentOptimizer(lr),
+        callbacks=Callbacks([
+            StatPrinter(),
+            ModelSaver(),
+            ScheduledHyperParamSetter('learning_rate', [(10, 1e-5), (20, 1e-6)])
+        ]),
+        model=model(),
+        step_per_epoch=step_per_epoch,
+        max_epoch=20,
+    )
+
+
+def visualize(modelpath, img_name="output"):
+    pred = OfflinePredictor(PredictConfig(
+        session_init=get_model_loader(modelpath),
+        model=EmbeddingModel(),
+        input_names=['input'],
+        output_names=['emb']))
+
+    NUM_BATCHES = 6
+    BATCH_SIZE = 128
+    images = np.zeros((BATCH_SIZE * NUM_BATCHES, 28, 28))  # the used digits
+    embed = np.zeros((BATCH_SIZE * NUM_BATCHES, 2))  # the actual embeddings in 2-d
+
+    # get only the embedding model data (MNIST test)
+    ds = EmbeddingModel.get_data()
+    ds.reset_state()
+
+    offset = 0
+    for digit, label in ds.get_data():
+        # TODO: investigate why [0][0]
+        prediction = pred([digit])[0][0]
```

Comment:
pred() return a list of outputs. You only want 1 output tensor so it's a list of length 1.
It's `[output tensor id][batch entry]`.